src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.cs

```diff
@@ -183,56 +219,98 @@ protected override void Dispose(bool disposing)
             }
 
             // Create RequestCompletionSource object and save current values of handler settings.
-            RequestCompletionSource state = new RequestCompletionSource();
-            state.CancellationToken = cancellationToken;
-            state.RequestMessage = request;
-            state.Handler = this;
+            RequestCompletionSource state = new RequestCompletionSource
+            {
+                CancellationToken = cancellationToken,
+                RequestMessage = request,
+            };
 
-            Task.Factory.StartNew(
-                s =>
-                {
-                    var rcs = (RequestCompletionSource)s;
-                    rcs.Handler.StartRequest(rcs);
-                },
-                state,
-                CancellationToken.None,
-                TaskCreationOptions.DenyChildAttach,
-                TaskScheduler.Default);
+            BeginRequest(state);
 
             return state.Task;
         }
 
         #region Private methods
 
-        private void StartRequest(RequestCompletionSource state)
+        private async void BeginRequest(RequestCompletionSource state)
         {
-            HttpResponseMessage responseMessage = null;
-            SafeCurlHandle requestHandle = null;
+            SafeCurlHandle requestHandle = new SafeCurlHandle();
+            GCHandle stateHandle = new GCHandle();
+            bool needCleanup = false;
 
-            if (state.CancellationToken.IsCancellationRequested)
+            try
             {
-                state.TrySetCanceled();
-                return;
-            }
+                // Prepare context objects
+                state.ResponseMessage = new CurlResponseMessage(state.RequestMessage);
+                stateHandle = GCHandle.Alloc(state);
+                requestHandle = CreateRequestHandle(state, stateHandle);
+                state.RequestHandle = requestHandle;
+                needCleanup = true;
 
-            var cancellationTokenRegistration = state.CancellationToken.Register(x => ((RequestCompletionSource) x).TrySetCanceled(), state);
+                if (state.CancellationToken.IsCancellationRequested)
```

Comment:
I'm not clear on why this cancellation check is necessary either.  The one below makes sense, because we will have just performed a potentially long operation and we want to poll cancellation again.  But this one, we just checked IsCancellationRequested prior to calling BeginRequest.